In [146]:
import pandas as pd


### Ejercicio 1

In [147]:
df= pd.read_csv(r"C:\Users\dpino\Documents\DATA SCIENCE\Jupyter\BOOTCAMP\T-3.2\archive\DelayedFlights.csv")
pd.set_option('display.max_columns', 500)
df.head(15)

Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
0            0  2008      1           3          4   2003.0        1955   
1            1  2008      1           3          4    754.0         735   
2            2  2008      1           3          4    628.0         620   
3            4  2008      1           3          4   1829.0        1755   
4            5  2008      1           3          4   1940.0        1915   
5            6  2008      1           3          4   1937.0        1830   
6           10  2008      1           3          4    706.0         700   
7           11  2008      1           3          4   1644.0        1510   
8           15  2008      1           3          4   1029.0        1020   
9           16  2008      1           3          4   1452.0        1425   
10          17  2008      1           3          4    754.0         745   
11          18  2008      1           3          4   1323.0        1255   
12          19  2008      1           3          4   1416.0        1325   
13          21  2008      1           3          4   1657.0        1625   
14          22  2008      1           3          4   1900.0        1840   

    ArrTime  CRSArrTime UniqueCarrier  FlightNum TailNum  ActualElapsedTime  \
0    2211.0        2225            WN        335  N712SW              128.0   
1    1002.0        1000            WN       3231  N772SW              128.0   
2     804.0         750            WN        448  N428WN               96.0   
3    1959.0        1925            WN       3920  N464WN               90.0   
4    2121.0        2110            WN        378  N726SW              101.0   
5    2037.0        1940            WN        509  N763SW              240.0   
6     916.0         915            WN        100  N690SW              130.0   
7    1845.0        1725            WN       1333  N334SW              121.0   
8    1021.0        1010            WN       2272  N263WN               52.0   
9    1640.0        1625            WN        675  N286WN              228.0   
10    940.0         955            WN       1144  N778SW              226.0   
11   1526.0        1510            WN          4  N674AA              123.0   
12   1512.0        1435            WN         54  N643SW               56.0   
13   1754.0        1735            WN        623  N724SW               57.0   
14   1956.0        1950            WN        717  N786SW               56.0   

    CRSElapsedTime  AirTime  ArrDelay  DepDelay Origin Dest  Distance  TaxiIn  \
0            150.0    116.0     -14.0       8.0    IAD  TPA       810     4.0   
1            145.0    113.0       2.0      19.0    IAD  TPA       810     5.0   
2             90.0     76.0      14.0       8.0    IND  BWI       515     3.0   
3             90.0     77.0      34.0      34.0    IND  BWI       515     3.0   
4            115.0     87.0      11.0      25.0    IND  JAX       688     4.0   
5            250.0    230.0      57.0      67.0    IND  LAS      1591     3.0   
6            135.0    106.0       1.0       6.0    IND  MCO       828     5.0   
7            135.0    107.0      80.0      94.0    IND  MCO       828     6.0   
8             50.0     37.0      11.0       9.0    IND  MDW       162     6.0   
9            240.0    213.0      15.0      27.0    IND  PHX      1489     7.0   
10           250.0    205.0     -15.0       9.0    IND  PHX      1489     5.0   
11           135.0    110.0      16.0      28.0    IND  TPA       838     4.0   
12            70.0     49.0      37.0      51.0    ISP  BWI       220     2.0   
13            70.0     47.0      19.0      32.0    ISP  BWI       220     5.0   
14            70.0     49.0       6.0      20.0    ISP  BWI       220     2.0   

    TaxiOut  Cancelled CancellationCode  Diverted  CarrierDelay  WeatherDelay  \
0       8.0          0                N         0           NaN           NaN   
1      10.0          0                N         0           NaN           NaN  

El dataframe contienen 30 columnas, que permiten saber fechas ( año, mes y dia ) sobre vuelos aereos. tambien nos brinda informacion exacta por vuelo, salida del vuelo, horario computarizado de salida de vuelo, hora de llegada, informacion de la compañia por cada vuelo, maniobras del avion en tierra, cancelaciones, codigos de calcelacion, desviacion del vuelo y ademas nos clasifica los motivos de retraso de vuelos ( Operadores, clima, NAS, Security, Retraso acumulado ).

Las columnas "DepTime", "CRSDepTime", "	ArrTime", "CRSArrTime", contienen una hora especifica.

El resto de columnas contienen datos en minutos. 

las columnas que clasifican los retrasos, contienen datos Nan, por que dependiendo el tipo de retraso se vera reflejado en la columna ArrDelay, una relacion encontrada es que cuando haya un registro en la columna ( LateAircraftDelay ) habra un registro en la columna (CarrierDelay), esto se suma y aparecera el total en ArrDelay.

CarrierDelay= Retraso por el operador del avion.

LateAircraftDelay= Retraso por que el avion ya venia de un vuelo anterior con retraso. 

taxiIn= Rodaje del avion desde la pista hasta el lugar de estacionamiento.

taxiout= Rodaje del avion hacia la pista, para iniciar vuelo.

SecurityDelaY = Retraso por cuestiones de seguridad.

NASDelay = Retrasos causados por el National Airspace System.

WeatherDelay = Retraso por clima

Tambien se identifica que:

AirTime + taxiIn + taxiOut= ActualElapsedTime (minutos), duracion del vuelo.

DepTime (dato de hora) - ArrTime (dato de hora) = ActualElapsedTime (minutos), duracion del vuelo.

DepTime (hora dada) + ActualElapsedTime (minutos)= ArrTime (hora de llegada real).

ArrTime (minutos,real) - ActualElapsedTime (minutos) = DepTime (hora dde salida real).

AirTime + ArrDelay + DepDelay + taxiIn + taxiOut= CRSElapsedTime (minutos) (computarizado).

DepTime - CRSDepTime = DepDelay (retraso que sufrio el vuelo a  su salida).

ArrTime - CRSArrTime = ArrDelay (retraso que sufrio el vuelo a  su llegada).

CRSArrTime - ArrDelay = ArrTime (Hora de llegada del vuelo).


In [148]:
df.isnull().sum()

Unnamed: 0                0
Year                      0
Month                     0
DayofMonth                0
DayOfWeek                 0
DepTime                   0
CRSDepTime                0
ArrTime                7110
CRSArrTime                0
UniqueCarrier             0
FlightNum                 0
TailNum                   5
ActualElapsedTime      8387
CRSElapsedTime          198
AirTime                8387
ArrDelay               8387
DepDelay                  0
Origin                    0
Dest                      0
Distance                  0
TaxiIn                 7110
TaxiOut                 455
Cancelled                 0
CancellationCode          0
Diverted                  0
CarrierDelay         689270
WeatherDelay         689270
NASDelay             689270
SecurityDelay        689270
LateAircraftDelay    689270
dtype: int64

Podemos ver, que faltan algunos datos en columnas (Nan),  por lo cual debemos convertir de dastos faltante a "0", pues las columnas aunque tengan datos faltantes son importantes para su analisis 

In [149]:
df1=df.where( df.notna(),0)
df1

Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
0                 0  2008      1           3          4   2003.0        1955   
1                 1  2008      1           3          4    754.0         735   
2                 2  2008      1           3          4    628.0         620   
3                 4  2008      1           3          4   1829.0        1755   
4                 5  2008      1           3          4   1940.0        1915   
...             ...   ...    ...         ...        ...      ...         ...   
1936753     7009710  2008     12          13          6   1250.0        1220   
1936754     7009717  2008     12          13          6    657.0         600   
1936755     7009718  2008     12          13          6   1007.0         847   
1936756     7009726  2008     12          13          6   1251.0        1240   
1936757     7009727  2008     12          13          6   1110.0        1103   

         ArrTime  CRSArrTime UniqueCarrier  FlightNum TailNum  \
0         2211.0        2225            WN        335  N712SW   
1         1002.0        1000            WN       3231  N772SW   
2          804.0         750            WN        448  N428WN   
3         1959.0        1925            WN       3920  N464WN   
4         2121.0        2110            WN        378  N726SW   
...          ...         ...           ...        ...     ...   
1936753   1617.0        1552            DL       1621  N938DL   
1936754    904.0         749            DL       1631  N3743H   
1936755   1149.0        1010            DL       1631  N909DA   
1936756   1446.0        1437            DL       1639  N646DL   
1936757   1413.0        1418            DL       1641  N908DL   

         ActualElapsedTime  CRSElapsedTime  AirTime  ArrDelay  DepDelay  \
0                    128.0           150.0    116.0     -14.0       8.0   
1                    128.0           145.0    113.0       2.0      19.0   
2                     96.0            90.0     76.0      14.0       8.0   
3                     90.0            90.0     77.0      34.0      34.0   
4                    101.0           115.0     87.0      11.0      25.0   
...                    ...             ...      ...       ...       ...   
1936753              147.0           152.0    120.0      25.0      30.0   
1936754              127.0           109.0     78.0      75.0      57.0   
1936755              162.0           143.0    122.0      99.0      80.0   
1936756              115.0           117.0     89.0       9.0      11.0   
1936757              123.0           135.0    104.0      -5.0       7.0   

        Origin Dest  Distance  TaxiIn  TaxiOut  Cancelled CancellationCode  \
0          IAD  TPA       810     4.0      8.0          0                N   
1          IAD  TPA       810     5.0     10.0          0                N   
2          IND  BWI       515     3.0     17.0          0                N   
3          IND  BWI       515     3.0     10.0          0                N   
4          IND  JAX       688     4.0     10.0          0                N   
...        ...  ...       ...     ...      ...        ...              ...   
1936753    MSP  ATL       906     9.0     18.0          0                N   
1936754    RIC  ATL       481    15.0     34.0          0                N   
1936755    ATL  IAH       689     8.0     32.0          0                N   
1936756    IAD  ATL       533    13.0     13.0          0                N   
1936757    SAT  ATL       874     8.0     11.0          0                N   

         Diverted  CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  \
0               0           0.0           0.0       0.0            0.0   
1               0           0.0           0.0       0.0            0.0   
2               0           0.0           0.0       0.0            0.0   
3               0           2.0           0.0       0.0            0.0   
4               0           0.0           0.0       0.0    

Los campos Nan, fueron reemplazados con "ceros"

In [150]:
df1.describe().T

count          mean           std     min        25%  \
Unnamed: 0         1936758.0  3.341651e+06  2.066065e+06     0.0  1517452.5   
Year               1936758.0  2.008000e+03  0.000000e+00  2008.0     2008.0   
Month              1936758.0  6.111106e+00  3.482546e+00     1.0        3.0   
DayofMonth         1936758.0  1.575347e+01  8.776272e+00     1.0        8.0   
DayOfWeek          1936758.0  3.984827e+00  1.995966e+00     1.0        2.0   
DepTime            1936758.0  1.518534e+03  4.504853e+02     1.0     1203.0   
CRSDepTime         1936758.0  1.467473e+03  4.247668e+02     0.0     1135.0   
ArrTime            1936758.0  1.604230e+03  5.557685e+02     0.0     1313.0   
CRSArrTime         1936758.0  1.634225e+03  4.646347e+02     0.0     1325.0   
FlightNum          1936758.0  2.184263e+03  1.944702e+03     1.0      610.0   
ActualElapsedTime  1936758.0  1.327286e+02  7.243471e+01     0.0       80.0   
CRSElapsedTime     1936758.0  1.342890e+02  7.135071e+01   -25.0       82.0   
AirTime            1936758.0  1.078083e+02  6.886184e+01     0.0       58.0   
ArrDelay           1936758.0  4.201714e+01  5.672935e+01  -109.0        9.0   
DepDelay           1936758.0  4.318518e+01  5.340250e+01     6.0       12.0   
Distance           1936758.0  7.656862e+02  5.744797e+02    11.0      338.0   
TaxiIn             1936758.0  6.787964e+00  5.280008e+00     0.0        4.0   
TaxiOut            1936758.0  1.822792e+01  1.433957e+01     0.0       10.0   
Cancelled          1936758.0  3.268348e-04  1.807562e-02     0.0        0.0   
Diverted           1936758.0  4.003598e-03  6.314722e-02     0.0        0.0   
CarrierDelay       1936758.0  1.235367e+01  3.613493e+01     0.0        0.0   
WeatherDelay       1936758.0  2.385512e+00  1.734036e+01     0.0        0.0   
NASDelay           1936758.0  9.675607e+00  2.808958e+01     0.0        0.0   
SecurityDelay      1936758.0  5.805836e-02  1.623934e+00     0.0        0.0   
LateAircraftDelay  1936758.0  1.629374e+01  3.585904e+01     0.0        0.0   

                         50%         75%        max  
Unnamed: 0         3242558.0  4972466.75  7009727.0  
Year                  2008.0     2008.00     2008.0  
Month                    6.0        9.00       12.0  
DayofMonth              16.0       23.00       31.0  
DayOfWeek                4.0        6.00        7.0  
DepTime               1545.0     1900.00     2400.0  
CRSDepTime            1510.0     1815.00     2359.0  
ArrTime               1714.0     2030.00     2400.0  
CRSArrTime            1705.0     2014.00     2400.0  
FlightNum             1543.0     3422.00     9742.0  
ActualElapsedTime      116.0      165.00     1114.0  
CRSElapsedTime         116.0      165.00      660.0  
AirTime                 90.0      137.00     1091.0  
ArrDelay                24.0       55.00     2461.0  
DepDelay                24.0       53.00     2467.0  
Distance               606.0      998.00     4962.0  
TaxiIn                   5.0        8.00      240.0  
TaxiOut                 14.0       21.00      422.0  
Cancelled                0.0        0.00        1.0  
Diverted                 0.0        0.00        1.0  
CarrierDelay             0.0       10.00     2436.0  
WeatherDelay             0.0        0.00     1352.0  
NASDelay                 0.0        6.00     1357.0  
SecurityDelay            0.0        0.00      392.0  
LateAircraftDelay        0.0       18.00     1316.0

Con df1.describe().T, Pandas, nos muestra informacion estadistica de las columnas numericas, que nos sirve para saber por ejemplo, si una columna tiene datos unicos como es el caso de "Year", "Cancelled", "Diverted", dado que la desviacion estandard de estas columnas es igual a 0. por lo cual puedo decir que no es una informacion relevante.

In [151]:
df2=df1.drop(["Diverted", "Cancelled", "Year", "Unnamed: 0", "CancellationCode"], axis=1)
df2

Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
0            1           3          4   2003.0        1955   2211.0   
1            1           3          4    754.0         735   1002.0   
2            1           3          4    628.0         620    804.0   
3            1           3          4   1829.0        1755   1959.0   
4            1           3          4   1940.0        1915   2121.0   
...        ...         ...        ...      ...         ...      ...   
1936753     12          13          6   1250.0        1220   1617.0   
1936754     12          13          6    657.0         600    904.0   
1936755     12          13          6   1007.0         847   1149.0   
1936756     12          13          6   1251.0        1240   1446.0   
1936757     12          13          6   1110.0        1103   1413.0   

         CRSArrTime UniqueCarrier  FlightNum TailNum  ActualElapsedTime  \
0              2225            WN        335  N712SW              128.0   
1              1000            WN       3231  N772SW              128.0   
2               750            WN        448  N428WN               96.0   
3              1925            WN       3920  N464WN               90.0   
4              2110            WN        378  N726SW              101.0   
...             ...           ...        ...     ...                ...   
1936753        1552            DL       1621  N938DL              147.0   
1936754         749            DL       1631  N3743H              127.0   
1936755        1010            DL       1631  N909DA              162.0   
1936756        1437            DL       1639  N646DL              115.0   
1936757        1418            DL       1641  N908DL              123.0   

         CRSElapsedTime  AirTime  ArrDelay  DepDelay Origin Dest  Distance  \
0                 150.0    116.0     -14.0       8.0    IAD  TPA       810   
1                 145.0    113.0       2.0      19.0    IAD  TPA       810   
2                  90.0     76.0      14.0       8.0    IND  BWI       515   
3                  90.0     77.0      34.0      34.0    IND  BWI       515   
4                 115.0     87.0      11.0      25.0    IND  JAX       688   
...                 ...      ...       ...       ...    ...  ...       ...   
1936753           152.0    120.0      25.0      30.0    MSP  ATL       906   
1936754           109.0     78.0      75.0      57.0    RIC  ATL       481   
1936755           143.0    122.0      99.0      80.0    ATL  IAH       689   
1936756           117.0     89.0       9.0      11.0    IAD  ATL       533   
1936757           135.0    104.0      -5.0       7.0    SAT  ATL       874   

         TaxiIn  TaxiOut  CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  \
0           4.0      8.0           0.0           0.0       0.0            0.0   
1           5.0     10.0           0.0           0.0       0.0            0.0   
2           3.0     17.0           0.0           0.0       0.0            0.0   
3           3.0     10.0           2.0           0.0       0.0            0.0   
4           4.0     10.0           0.0           0.0       0.0            0.0   
...         ...      ...           ...           ...       ...            ...   
1936753     9.0     18.0           3.0           0.0       0.0            0.0   
1936754    15.0     34.0           0.0          57.0      18.0            0.0   
1936755     8.0     32.0           1.0           0.0      19.0            0.0   
1936756    13.0     13.0           0.0           0.0       0.0            0.0   
1936757     8.0     11.0           0.0           0.0       0.0            0.0   

         LateAircraftDelay  
0                      0.0  
1                      0.0  
2                      0.0  
3                     32.0  
4                      0.0  
...                    ...  
1936753               22.0  
1936754                0.0  
1936755               79.0  
1936756                0.0  
1936757                0.0

El dataframe contiene 25 columnas que considero son importantes, pues se relacionan numericamente entre ellas o tienen informacion relevante pasra identificar los vuelos. 

In [152]:
df3 = df2.sample(n=200000)
df3

Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
1594677     10           1          3   1747.0        1620   1828.0   
138580       1          31          4   1656.0        1315   1811.0   
846307       5          14          3   2124.0        2025     40.0   
1018543      6           6          5   1550.0        1455   1759.0   
806132       5           1          4   1422.0        1350   1935.0   
...        ...         ...        ...      ...         ...      ...   
1724501     11          30          7    755.0         745   1130.0   
1210395      7          14          1   1430.0        1315   1534.0   
1591188     10           2          4   1530.0        1439   1605.0   
213601       2          24          7   2143.0        2040   2324.0   
175618       1          11          5   1715.0        1545   1942.0   

         CRSArrTime UniqueCarrier  FlightNum TailNum  ActualElapsedTime  \
1594677        1710            MQ       3893  N834MQ               41.0   
138580         1425            9E       2005  91469E              135.0   
846307         2310            AA        345  N589AA              196.0   
1018543        1630            MQ       4419  N839AE              189.0   
806132         1915            WN       2008  N298WN              193.0   
...             ...           ...        ...     ...                ...   
1724501        1112            CO        260  N12318              275.0   
1210395        1425            MQ       4818  N817AE               64.0   
1591188        1510            HA        160  N476HA               35.0   
213601         2230            WN       1286  N278WN              101.0   
175618         1820            CO       1156  N16642              147.0   

         CRSElapsedTime  AirTime  ArrDelay  DepDelay Origin Dest  Distance  \
1594677            50.0     29.0      78.0      87.0    DFW  ABI       158   
138580            130.0    113.0     226.0     221.0    ATL  IAH       689   
846307            165.0    166.0      90.0      59.0    ORD  SAT      1041   
1018543           155.0    120.0      89.0      55.0    HPN  ORD       738   
806132            205.0    180.0      20.0      32.0    OAK  MCI      1489   
...                 ...      ...       ...       ...    ...  ...       ...   
1724501           267.0    241.0      18.0      10.0    EWR  SAT      1569   
1210395            70.0     44.0      69.0      75.0    BOS  LGA       185   
1591188            31.0     23.0      55.0      51.0    OGG  KOA        84   
213601            110.0     88.0      54.0      63.0    SMF  SEA       605   
175618            155.0     96.0      82.0      90.0    ATL  EWR       745   

         TaxiIn  TaxiOut  CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  \
1594677     1.0     11.0          78.0           0.0       0.0            0.0   
138580      4.0     18.0           0.0           0.0       5.0            0.0   
846307      5.0     25.0          23.0           0.0      31.0            0.0   
1018543    23.0     46.0           0.0           0.0      85.0            0.0   
806132      4.0      9.0           0.0           0.0       0.0            0.0   
...         ...      ...           ...           ...       ...            ...   
1724501     4.0     30.0          10.0           0.0       8.0            0.0   
1210395     8.0     12.0           0.0           0.0      69.0            0.0   
1591188     6.0      6.0           4.0           0.0       0.0            0.0   
213601      6.0      7.0           0.0           0.0       0.0            0.0   
175618      9.0     42.0           0.0           0.0      82.0            0.0   

         LateAircraftDelay  
1594677                0.0  
138580               221.0  
846307                36.0  
1018543                4.0  
806132                20.0  
...                    ...  
1724501                0.0  
1210395                0.0  
1591188               51.0  
213601                54.0  
175618                 0.0

### Ejercicio 2

#### Resume estadísticamente el dataset y las columnas de interés. Haz un análisis estadístico de lo que consideres relevante.

In [153]:
df3.describe().T

count         mean          std   min     25%     50%  \
Month              200000.0     6.106210     3.480655   1.0     3.0     6.0   
DayofMonth         200000.0    15.770760     8.761502   1.0     8.0    16.0   
DayOfWeek          200000.0     3.983645     1.997398   1.0     2.0     4.0   
DepTime            200000.0  1517.447470   450.562798   1.0  1202.0  1544.0   
CRSDepTime         200000.0  1466.133585   425.246774   1.0  1132.0  1509.0   
ArrTime            200000.0  1603.901195   555.033678   0.0  1312.0  1713.0   
CRSArrTime         200000.0  1632.869045   464.876254   0.0  1325.0  1705.0   
FlightNum          200000.0  2185.818690  1941.370361   1.0   611.0  1549.0   
ActualElapsedTime  200000.0   132.791760    72.191221   0.0    80.0   116.0   
CRSElapsedTime     200000.0   134.327915    71.115139 -12.0    82.0   116.0   
AirTime            200000.0   107.823325    68.598220   0.0    58.0    90.0   
ArrDelay           200000.0    41.980895    56.880662 -70.0     9.0    24.0   
DepDelay           200000.0    43.153485    53.525045   6.0    12.0    24.0   
Distance           200000.0   765.845790   572.453477  31.0   341.0   606.0   
TaxiIn             200000.0     6.809675     5.336412   0.0     4.0     6.0   
TaxiOut            200000.0    18.251530    14.424279   0.0    10.0    14.0   
CarrierDelay       200000.0    12.416860    36.068028   0.0     0.0     0.0   
WeatherDelay       200000.0     2.361925    17.572346   0.0     0.0     0.0   
NASDelay           200000.0     9.691355    28.292423   0.0     0.0     0.0   
SecurityDelay      200000.0     0.061260     1.706491   0.0     0.0     0.0   
LateAircraftDelay  200000.0    16.200755    35.889375   0.0     0.0     0.0   

                      75%     max  
Month                 9.0    12.0  
DayofMonth           23.0    31.0  
DayOfWeek             6.0     7.0  
DepTime            1900.0  2400.0  
CRSDepTime         1815.0  2359.0  
ArrTime            2030.0  2400.0  
CRSArrTime         2014.0  2400.0  
FlightNum          3420.0  9742.0  
ActualElapsedTime   165.0   702.0  
CRSElapsedTime      165.0   660.0  
AirTime             137.0   654.0  
ArrDelay             55.0  1417.0  
DepDelay             53.0  1428.0  
Distance            998.0  4962.0  
TaxiIn                8.0   240.0  
TaxiOut              21.0   364.0  
CarrierDelay         10.0  1172.0  
WeatherDelay          0.0  1352.0  
NASDelay              6.0  1337.0  
SecurityDelay         0.0   254.0  
LateAircraftDelay    18.0  1236.0

En este analisis estadistico, se muestra un dataset de 200.000, registros, nos muestra: mean,std, min, 25%, 50%, 75%, max.
donde podemos observar que:

DepTime= El promedio de hora en que salieron los aviones, fue a las 13:19 h, frente a las 14:68 h (CRSDepTime), computarizado, lo cual indica que hubo un retraso en las salidas, la mayoria de aviones, tuvo salida antes de las 19:01 (DepTime), sin embargo se esperaba (CRSDepTime) que la mayoria hubiese salido antes de las 18:19 h. 

ArrTime= El promedio de hora de llegada de los vuelos, fue a las  16:03 h, frente a las 16:35, que se tenia proyectado, lo cual indica que en promedio los vuelos llegaron antes de tiempo aunque tuvieron un retraso en su salida, la mayoria de llegadas, fueron a las 20:30, frente a las proyectadas 20:15 h.

ActualElapsedTime= En promedio los vuelos tuvieron una duraciòn de 132 min y la mayoria tuvo una duracion de 165 min igual a lo proyectado (CRSElapsedTime)

AirTime= los vuelos tuvieron un promedio de tiempo en el aire de de 107.8 min, el vuelo mas largo fue de 624 min. el 75% de los vuelos tuvo una diracion de 137 min. 

ArrDelay= Los vuelos tuvieron un promedio de retraso de llegada de 41.9 min,  el maximo de retraso fue de 1308 min. y el minimo de -66, lo cual indica que los vuelos llegaron antes de tiempo.

DepDelay= Los vuelos tuvieron un promedio de retraso en la salida de 43 min, el retraso minimo en la salida fue de 6 min, mientras que el maximo fue de 1395 min, la mayoria de salidas tuvieron un retraso de 53 min.

Distance= Los vuelos recorrieron una distancia en promedio de 765 km, la minima distancia recorrida por un vuelo fue de 11 km, mientras que la maxima due de 4.962 km., la mayoria de aviones recorrio una distancia de 998 km.

CarrierDelay= El retraso maximo por los operadores fue de 1382 min

WeatherDelay= El lretraso maximo por tiempo fue de 1225 min.

SecurityDelay= El retraso maximo por seguridad fue de 284 min.


#### Encuentra cuántos datos faltantes hay por columna.

In [154]:
df3.isnull().sum()

Month                0
DayofMonth           0
DayOfWeek            0
DepTime              0
CRSDepTime           0
ArrTime              0
CRSArrTime           0
UniqueCarrier        0
FlightNum            0
TailNum              0
ActualElapsedTime    0
CRSElapsedTime       0
AirTime              0
ArrDelay             0
DepDelay             0
Origin               0
Dest                 0
Distance             0
TaxiIn               0
TaxiOut              0
CarrierDelay         0
WeatherDelay         0
NASDelay             0
SecurityDelay        0
LateAircraftDelay    0
dtype: int64

#### Crea nuevas columnas (velocidad media del vuelo, si ha llegado tarde o no...).

In [155]:
df3["Vel media vuelo"]= round(df3["Distance"]/df3["AirTime"])
df3

Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
1594677     10           1          3   1747.0        1620   1828.0   
138580       1          31          4   1656.0        1315   1811.0   
846307       5          14          3   2124.0        2025     40.0   
1018543      6           6          5   1550.0        1455   1759.0   
806132       5           1          4   1422.0        1350   1935.0   
...        ...         ...        ...      ...         ...      ...   
1724501     11          30          7    755.0         745   1130.0   
1210395      7          14          1   1430.0        1315   1534.0   
1591188     10           2          4   1530.0        1439   1605.0   
213601       2          24          7   2143.0        2040   2324.0   
175618       1          11          5   1715.0        1545   1942.0   

         CRSArrTime UniqueCarrier  FlightNum TailNum  ActualElapsedTime  \
1594677        1710            MQ       3893  N834MQ               41.0   
138580         1425            9E       2005  91469E              135.0   
846307         2310            AA        345  N589AA              196.0   
1018543        1630            MQ       4419  N839AE              189.0   
806132         1915            WN       2008  N298WN              193.0   
...             ...           ...        ...     ...                ...   
1724501        1112            CO        260  N12318              275.0   
1210395        1425            MQ       4818  N817AE               64.0   
1591188        1510            HA        160  N476HA               35.0   
213601         2230            WN       1286  N278WN              101.0   
175618         1820            CO       1156  N16642              147.0   

         CRSElapsedTime  AirTime  ArrDelay  DepDelay Origin Dest  Distance  \
1594677            50.0     29.0      78.0      87.0    DFW  ABI       158   
138580            130.0    113.0     226.0     221.0    ATL  IAH       689   
846307            165.0    166.0      90.0      59.0    ORD  SAT      1041   
1018543           155.0    120.0      89.0      55.0    HPN  ORD       738   
806132            205.0    180.0      20.0      32.0    OAK  MCI      1489   
...                 ...      ...       ...       ...    ...  ...       ...   
1724501           267.0    241.0      18.0      10.0    EWR  SAT      1569   
1210395            70.0     44.0      69.0      75.0    BOS  LGA       185   
1591188            31.0     23.0      55.0      51.0    OGG  KOA        84   
213601            110.0     88.0      54.0      63.0    SMF  SEA       605   
175618            155.0     96.0      82.0      90.0    ATL  EWR       745   

         TaxiIn  TaxiOut  CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  \
1594677     1.0     11.0          78.0           0.0       0.0            0.0   
138580      4.0     18.0           0.0           0.0       5.0            0.0   
846307      5.0     25.0          23.0           0.0      31.0            0.0   
1018543    23.0     46.0           0.0           0.0      85.0            0.0   
806132      4.0      9.0           0.0           0.0       0.0            0.0   
...         ...      ...           ...           ...       ...            ...   
1724501     4.0     30.0          10.0           0.0       8.0            0.0   
1210395     8.0     12.0           0.0           0.0      69.0            0.0   
1591188     6.0      6.0           4.0           0.0       0.0            0.0   
213601      6.0      7.0           0.0           0.0       0.0            0.0   
175618      9.0     42.0           0.0           0.0      82.0            0.0   

         LateAircraftDelay  Vel media vuelo  
1594677                0.0              5.0  
138580               221.0              6.0  
846307                36.0              6.0  
1018543                4.0              6.0  
806132                20.0              8.0  
...                    ...              ...  
1724501                

Se ha creado una nueva columna Vel. promedio de vuelo, donde interactuan las columnas de "Distance" y "Airtime", se puede 
observar que algunas filas muestran resultado"inf", esto es debido a que alguna de las dos columnas tienen un registro en "0" 

In [156]:
df3['ClassArrDelay'] = df3['ArrDelay'].apply(lambda x: 'Tarde' if x >= 1.0 else 'A tiempo')
df3

Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
1594677     10           1          3   1747.0        1620   1828.0   
138580       1          31          4   1656.0        1315   1811.0   
846307       5          14          3   2124.0        2025     40.0   
1018543      6           6          5   1550.0        1455   1759.0   
806132       5           1          4   1422.0        1350   1935.0   
...        ...         ...        ...      ...         ...      ...   
1724501     11          30          7    755.0         745   1130.0   
1210395      7          14          1   1430.0        1315   1534.0   
1591188     10           2          4   1530.0        1439   1605.0   
213601       2          24          7   2143.0        2040   2324.0   
175618       1          11          5   1715.0        1545   1942.0   

         CRSArrTime UniqueCarrier  FlightNum TailNum  ActualElapsedTime  \
1594677        1710            MQ       3893  N834MQ               41.0   
138580         1425            9E       2005  91469E              135.0   
846307         2310            AA        345  N589AA              196.0   
1018543        1630            MQ       4419  N839AE              189.0   
806132         1915            WN       2008  N298WN              193.0   
...             ...           ...        ...     ...                ...   
1724501        1112            CO        260  N12318              275.0   
1210395        1425            MQ       4818  N817AE               64.0   
1591188        1510            HA        160  N476HA               35.0   
213601         2230            WN       1286  N278WN              101.0   
175618         1820            CO       1156  N16642              147.0   

         CRSElapsedTime  AirTime  ArrDelay  DepDelay Origin Dest  Distance  \
1594677            50.0     29.0      78.0      87.0    DFW  ABI       158   
138580            130.0    113.0     226.0     221.0    ATL  IAH       689   
846307            165.0    166.0      90.0      59.0    ORD  SAT      1041   
1018543           155.0    120.0      89.0      55.0    HPN  ORD       738   
806132            205.0    180.0      20.0      32.0    OAK  MCI      1489   
...                 ...      ...       ...       ...    ...  ...       ...   
1724501           267.0    241.0      18.0      10.0    EWR  SAT      1569   
1210395            70.0     44.0      69.0      75.0    BOS  LGA       185   
1591188            31.0     23.0      55.0      51.0    OGG  KOA        84   
213601            110.0     88.0      54.0      63.0    SMF  SEA       605   
175618            155.0     96.0      82.0      90.0    ATL  EWR       745   

         TaxiIn  TaxiOut  CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  \
1594677     1.0     11.0          78.0           0.0       0.0            0.0   
138580      4.0     18.0           0.0           0.0       5.0            0.0   
846307      5.0     25.0          23.0           0.0      31.0            0.0   
1018543    23.0     46.0           0.0           0.0      85.0            0.0   
806132      4.0      9.0           0.0           0.0       0.0            0.0   
...         ...      ...           ...           ...       ...            ...   
1724501     4.0     30.0          10.0           0.0       8.0            0.0   
1210395     8.0     12.0           0.0           0.0      69.0            0.0   
1591188     6.0      6.0           4.0           0.0       0.0            0.0   
213601      6.0      7.0           0.0           0.0       0.0            0.0   
175618      9.0     42.0           0.0           0.0      82.0            0.0   

         LateAircraftDelay  Vel media vuelo ClassArrDelay  
1594677                0.0              5.0         Tarde  
138580               221.0              6.0         Tarde  
846307                36.0              6.0         Tarde  
1018543                4.0              6.0         Tarde  
806132                20.0              8.0  

In [157]:
df3['WhyDelay'] = df3[['CarrierDelay', 'WeatherDelay', 'NASDelay', 'SecurityDelay', 'LateAircraftDelay']].apply(
    lambda row: 'undefined' if all(val == 0 for val in row) else row.idxmax(),
    axis=1)
df3

Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
1594677     10           1          3   1747.0        1620   1828.0   
138580       1          31          4   1656.0        1315   1811.0   
846307       5          14          3   2124.0        2025     40.0   
1018543      6           6          5   1550.0        1455   1759.0   
806132       5           1          4   1422.0        1350   1935.0   
...        ...         ...        ...      ...         ...      ...   
1724501     11          30          7    755.0         745   1130.0   
1210395      7          14          1   1430.0        1315   1534.0   
1591188     10           2          4   1530.0        1439   1605.0   
213601       2          24          7   2143.0        2040   2324.0   
175618       1          11          5   1715.0        1545   1942.0   

         CRSArrTime UniqueCarrier  FlightNum TailNum  ActualElapsedTime  \
1594677        1710            MQ       3893  N834MQ               41.0   
138580         1425            9E       2005  91469E              135.0   
846307         2310            AA        345  N589AA              196.0   
1018543        1630            MQ       4419  N839AE              189.0   
806132         1915            WN       2008  N298WN              193.0   
...             ...           ...        ...     ...                ...   
1724501        1112            CO        260  N12318              275.0   
1210395        1425            MQ       4818  N817AE               64.0   
1591188        1510            HA        160  N476HA               35.0   
213601         2230            WN       1286  N278WN              101.0   
175618         1820            CO       1156  N16642              147.0   

         CRSElapsedTime  AirTime  ArrDelay  DepDelay Origin Dest  Distance  \
1594677            50.0     29.0      78.0      87.0    DFW  ABI       158   
138580            130.0    113.0     226.0     221.0    ATL  IAH       689   
846307            165.0    166.0      90.0      59.0    ORD  SAT      1041   
1018543           155.0    120.0      89.0      55.0    HPN  ORD       738   
806132            205.0    180.0      20.0      32.0    OAK  MCI      1489   
...                 ...      ...       ...       ...    ...  ...       ...   
1724501           267.0    241.0      18.0      10.0    EWR  SAT      1569   
1210395            70.0     44.0      69.0      75.0    BOS  LGA       185   
1591188            31.0     23.0      55.0      51.0    OGG  KOA        84   
213601            110.0     88.0      54.0      63.0    SMF  SEA       605   
175618            155.0     96.0      82.0      90.0    ATL  EWR       745   

         TaxiIn  TaxiOut  CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  \
1594677     1.0     11.0          78.0           0.0       0.0            0.0   
138580      4.0     18.0           0.0           0.0       5.0            0.0   
846307      5.0     25.0          23.0           0.0      31.0            0.0   
1018543    23.0     46.0           0.0           0.0      85.0            0.0   
806132      4.0      9.0           0.0           0.0       0.0            0.0   
...         ...      ...           ...           ...       ...            ...   
1724501     4.0     30.0          10.0           0.0       8.0            0.0   
1210395     8.0     12.0           0.0           0.0      69.0            0.0   
1591188     6.0      6.0           4.0           0.0       0.0            0.0   
213601      6.0      7.0           0.0           0.0       0.0            0.0   
175618      9.0     42.0           0.0           0.0      82.0            0.0   

         LateAircraftDelay  Vel media vuelo ClassArrDelay           WhyDelay  
1594677                0.0              5.0         Tarde       CarrierDelay  
138580               221.0              6.0         Tarde  LateAircraftDelay  
846307                36.0              6.0         Tarde  LateAircraftDelay  
1018543                4.0   

En la columna creada "WhyDelay" queria mostrar el motivo del retraso, dado que existen 5 posibles motivos(5 columnas) sin embargo las que estan en 0.0 tienen otros motivos no especificados y los define como "undefined"

#### Haz una mesa de las aerolíneas con mayores retrasos acumulados.

In [158]:
restrasos_acumulados= df3.groupby(by = "UniqueCarrier")["ArrDelay"].sum().head(10)
restrasos_acumulados

UniqueCarrier
9E    253307.0
AA    909600.0
AQ      2763.0
AS    146348.0
B6    314377.0
CO    415806.0
DL    471940.0
EV    392629.0
F9     81257.0
FL    323464.0
Name: ArrDelay, dtype: float64

#### ¿Cuáles son los vuelos más largos?

In [159]:
df3.sort_values(by='ActualElapsedTime', ascending=False)[['ActualElapsedTime', 'FlightNum']].head(10)

ActualElapsedTime  FlightNum
1926347              702.0         15
570298               697.0       1273
556378               686.0         15
875499               681.0         15
556379               672.0         15
363637               666.0         15
871328               658.0         15
360576               653.0         15
1729952              653.0         15
556366               645.0         15

un mismo numero de vuelo, realizo los vuelos mas largos

#### ¿Y los más atrasados?

In [160]:
df3.sort_values(by='ArrDelay', ascending=False)[['ArrDelay', 'FlightNum']].head(10)

ArrDelay  FlightNum
1875741    1417.0       4333
1228543    1337.0       1073
1901407    1312.0       1138
1702698    1253.0       4508
1906664    1225.0       1790
545402     1182.0       2228
840195     1177.0       1509
322560     1169.0        821
1887360    1155.0       1639
533478     1142.0        744

#### Busca las rutas más largas

In [161]:
df3['Route'] = df3['Origin'] + '-' + df['Dest']
df3.sort_values(by='ActualElapsedTime', ascending=False)[['ActualElapsedTime', 'Route']].head(10)

ActualElapsedTime    Route
1926347              702.0  EWR-HNL
570298               697.0  ATL-HNL
556378               686.0  EWR-HNL
875499               681.0  EWR-HNL
556379               672.0  EWR-HNL
363637               666.0  EWR-HNL
871328               658.0  EWR-HNL
360576               653.0  EWR-HNL
1729952              653.0  EWR-HNL
556366               645.0  EWR-HNL

Concateno las columnas "Origen", "Dest" para crear una columna unica que sea "Route".

#### Las que acumulen más retrasos.

In [162]:
total_route = df3.groupby('Route')['ArrDelay'].sum().reset_index()
total_route.sort_values(by='ArrDelay', ascending=False).head(10)

Route  ArrDelay
2528  LAX-SFO   24008.0
3497  ORD-LGA   22074.0
2603  LGA-ORD   21590.0
1652  EWR-ORD   21051.0
214   ATL-LGA   20503.0
3466  ORD-EWR   20065.0
176   ATL-EWR   19913.0
845   CLT-EWR   17939.0
4301  SFO-LAX   17795.0
3210  MSP-ORD   17492.0

#### Aerolineas con distancias y rutas mas largas, contrastado con su puntualidad.

In [163]:
df3.sort_values(by='Distance', ascending=False)[['Distance', 'Route', "UniqueCarrier", "ClassArrDelay"]].head(10)

Distance    Route UniqueCarrier ClassArrDelay
871328       4962  EWR-HNL            CO         Tarde
872155       4962  EWR-HNL            CO         Tarde
875964       4962  EWR-HNL            CO         Tarde
365506       4962  EWR-HNL            CO      A tiempo
363637       4962  EWR-HNL            CO         Tarde
1420759      4962  EWR-HNL            CO         Tarde
1618504      4962  EWR-HNL            CO         Tarde
556378       4962  EWR-HNL            CO         Tarde
556366       4962  EWR-HNL            CO      A tiempo
556387       4962  EWR-HNL            CO         Tarde

las rutas con las distancias mas largas, con su respectivo operador, tambien quise ver el comportamiento de estas rutas respecto al ArrDelay si llego tarde o no. vemos que de esta muetras de 10, las que llegaron tarde fueron 5 y temprano 5 

In [164]:
df3

Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
1594677     10           1          3   1747.0        1620   1828.0   
138580       1          31          4   1656.0        1315   1811.0   
846307       5          14          3   2124.0        2025     40.0   
1018543      6           6          5   1550.0        1455   1759.0   
806132       5           1          4   1422.0        1350   1935.0   
...        ...         ...        ...      ...         ...      ...   
1724501     11          30          7    755.0         745   1130.0   
1210395      7          14          1   1430.0        1315   1534.0   
1591188     10           2          4   1530.0        1439   1605.0   
213601       2          24          7   2143.0        2040   2324.0   
175618       1          11          5   1715.0        1545   1942.0   

         CRSArrTime UniqueCarrier  FlightNum TailNum  ActualElapsedTime  \
1594677        1710            MQ       3893  N834MQ               41.0   
138580         1425            9E       2005  91469E              135.0   
846307         2310            AA        345  N589AA              196.0   
1018543        1630            MQ       4419  N839AE              189.0   
806132         1915            WN       2008  N298WN              193.0   
...             ...           ...        ...     ...                ...   
1724501        1112            CO        260  N12318              275.0   
1210395        1425            MQ       4818  N817AE               64.0   
1591188        1510            HA        160  N476HA               35.0   
213601         2230            WN       1286  N278WN              101.0   
175618         1820            CO       1156  N16642              147.0   

         CRSElapsedTime  AirTime  ArrDelay  DepDelay Origin Dest  Distance  \
1594677            50.0     29.0      78.0      87.0    DFW  ABI       158   
138580            130.0    113.0     226.0     221.0    ATL  IAH       689   
846307            165.0    166.0      90.0      59.0    ORD  SAT      1041   
1018543           155.0    120.0      89.0      55.0    HPN  ORD       738   
806132            205.0    180.0      20.0      32.0    OAK  MCI      1489   
...                 ...      ...       ...       ...    ...  ...       ...   
1724501           267.0    241.0      18.0      10.0    EWR  SAT      1569   
1210395            70.0     44.0      69.0      75.0    BOS  LGA       185   
1591188            31.0     23.0      55.0      51.0    OGG  KOA        84   
213601            110.0     88.0      54.0      63.0    SMF  SEA       605   
175618            155.0     96.0      82.0      90.0    ATL  EWR       745   

         TaxiIn  TaxiOut  CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  \
1594677     1.0     11.0          78.0           0.0       0.0            0.0   
138580      4.0     18.0           0.0           0.0       5.0            0.0   
846307      5.0     25.0          23.0           0.0      31.0            0.0   
1018543    23.0     46.0           0.0           0.0      85.0            0.0   
806132      4.0      9.0           0.0           0.0       0.0            0.0   
...         ...      ...           ...           ...       ...            ...   
1724501     4.0     30.0          10.0           0.0       8.0            0.0   
1210395     8.0     12.0           0.0           0.0      69.0            0.0   
1591188     6.0      6.0           4.0           0.0       0.0            0.0   
213601      6.0      7.0           0.0           0.0       0.0            0.0   
175618      9.0     42.0           0.0           0.0      82.0            0.0   

         LateAircraftDelay  Vel media vuelo ClassArrDelay           WhyDelay  \
1594677                0.0              5.0         Tarde       CarrierDelay   
138580               221.0              6.0         Tarde  LateAircraftDelay   
846307                36.0              6.0         Tarde  LateAircraftDelay   
1018543                4.

### Ejercicio 3

In [ ]:
df3.to_excel('Flight_Delay_Analysis', index=False)